### PR TITLE
[flutter_tools] always initialize the resident runner from dill

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Copy the output dill to a shared location `build/cache.dill` after running, and initialize new runners from this dill. Doesn't attempt to handle different build configurations or updates to dart, since this is a best effort only performance improvement.

Along with https://dart-review.googlesource.com/c/sdk/+/139313 , this change speeds up subsequent compiles of the flutter gallery (in web) from 14-16 seconds to 7-8 seconds.

https://github.com/flutter/flutter/issues/50616

### CODE REVIEW

**Review by jakemac53 — APPROVED** [review]

**Comment by zanderso** [comment]
> Doesn't attempt to handle different build configurations or updates to dart, since this is a best effort only performance improvement.

Will builds with different configurations be incorrect if a stale cached file exists?

**Comment by jonahwilliams** [comment]
> Will builds with different configurations be incorrect if a stale cached file exists?

No, they just won't benefit from the speed up. Tested by switching back and forth between a web and ios sim build.

**Comment by jonahwilliams** [comment]
The initializing logic handles certain classes of errors here: https://github.com/dart-lang/sdk/blob/3c4089da96f52a66165abb2e09487dd86a4ae600/pkg/front_end/lib/src/fasta/incremental_compiler.dart#L1009

I'll need to do more research to see if we can hit any of the unhandled paths with this change

**Comment by jonahwilliams** [comment]
Updated with a comment on how initialize-from-dill handles failures

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -75,6 +75,12 @@ class FlutterDevice {
     if (device.platformType == PlatformType.fuchsia) {
       targetModel = TargetModel.flutterRunner;
     }
+    // For both web and non-web platforms we initialize dill to/from
+    // a shared location for faster bootstrapping. If this process fails
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Maybe: 'this process' --> 'the compiler'.

**Review by zanderso — APPROVED** [review]
lgtm w/ small comment nit.

**Comment by jonahwilliams** [comment]
The dart sdk change rolled, landing

**Comment by jonahwilliams** [comment]
This caused a nice drop in the second startup time benchmarks

Screen Shot 2020-03-17 at 5 15 42 PM

**Pull request merged**